contracts/RandomBeaconHistory.cdc

```diff
@@ -0,0 +1,96 @@
+// 
+// For usage of randomness where result abortion is not an issue, it is recommended
+// to use the Cadence built-in function `revertibleRandom`, which is also based on 
+// the safe Random Beacon. 
+///
+/// Read the full FLIP here: https://github.com/onflow/flips/pull/123
+///
+access(all) contract RandomBeaconHistory {
+
+    /// The height at which the first source of randomness was recorded
+    access(contract) var initHeight: UInt64?
+    /// Sequence of random sources recorded by the Heartbeat, stored as an array over a mapping to reduce storage
+    access(contract) let randomSourceHistory: [[UInt8]]
+
+    /// The path of the Heartbeat resource in the deployment account
+    pub let HeartbeatStoragePath: StoragePath
+
+    /* --- Hearbeat --- */
+    //
+    /// The Heartbeat resource containing each block's source of randomness in sequence
+    ///
+    access(all) resource Heartbeat {
+
+        /// Callable by owner of the Heartbeat resource, Flow Service Account, records the provided random source
+        ///
+        /// @param randomSourceHistory The random source to record
+        ///
+        access(all) fun heartbeat(randomSourceHistory: [UInt8]) {
+
+            let currentBlockHeight = getCurrentBlock().height
+            if RandomBeaconHistory.initHeight == nil {
+                RandomBeaconHistory.initHeight = currentBlockHeight
+            }
+
+            RandomBeaconHistory.randomSourceHistory.append(randomSourceHistory)
+        }
+    }
+
+    /// Getter for the source of randomness at a given block height. Panics if the requested block height either
+    /// precedes or exceeds the recorded history. Note that a source of randomness for block n will not be accessible
+    /// until block n+1.
+    ///
+    /// @param atBlockHeight The block height at which to retrieve the source of randomness
+    /// @return The source of randomness at the given block height
+    ///
+    access(all) fun sourceOfRandomness(atBlockHeight: UInt64): [UInt8] {
+        pre {
+            self.initHeight != nil: "History has not yet been initialized"
+            atBlockHeight >= self.initHeight!: "Requested block height precedes recorded history"
+            atBlockHeight < getCurrentBlock().height: "Source of randomness not yet recorded"
+        }
+        let index: UInt64 = atBlockHeight - self.initHeight!
+
+        assert(
+            index >= 0 && index < UInt64(self.randomSourceHistory.length),
+            message: "Problem finding random source history index"
+        )
+
+        return self.randomSourceHistory[index]
+    }
+
+    /// Getter for the totality of recorded randomness source history
+    ///
+    /// @return An array of random sources, each source an array of UInt8
+    ///
+    access(all) view fun getRandomSourceHistory(): [[UInt8]] {
+        return self.initHeight != nil ? self.randomSourceHistory : panic("History has not yet been initialized")
+    }
```

Comment:
If I'm not mistaken returning `self.randomSourceHistory` will copy the underlying array. This will quickly render this function un-callable. 

If we really want something like this we can add bounds to the call:
```cadence
access(all) view fun getRandomSourceHistory(startHeight: Uint64, endHeight: Uint64): [[UInt8]] 
```

or model it as a paged result.
```cadence
getRandomSourceHistoryPage(page: UInt64, perPage: UInt): RandomSourceHistoryPage
//...
struct RandomSourceHistoryPage{
    perPage: UInt
    totalRecords: UInt
    startBlockHeight: UInt
   values: [[UInt8]] 
}
```

Reply:
> If I'm not mistaken returning self.randomSourceHistory will copy the underlying array. This will quickly render this function un-callable.

yeah this will grow really fast. 

btw after stable cadence we can return a reference probably. iirc modifications require entitlements after stable cadence.

Reply:
Very true. I think returning non-entitled reference will be most efficient come stable cadence. For now though, I think modeling a paged result would be consistent with other core contracts. Thanks for the callout, will update shortly